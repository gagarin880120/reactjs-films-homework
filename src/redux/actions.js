function searchResultsAction(results) {
  return {
      type: 'SEARCH',
      results: results
  };
}

function genresAction(genres) {
  return {
      type: 'GENRES',
      genres: genres
  };
}

export { searchResultsAction, genresAction };
