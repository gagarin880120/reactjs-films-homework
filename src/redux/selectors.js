import getGenresString from '../helpers/helpers';

const searchResultsSelector = (state) => {
  state.searchResults.map((v) => {
    const obj = v;
    obj.genres = getGenresString(v.genre_ids, state.genres);
    return { ...obj };
  });
  return state.searchResults;
};

const modalSelector = (state) => state.isModalOpen;

const trailerSelector = (state) => state.trailerURL;


export { searchResultsSelector, modalSelector, trailerSelector };
