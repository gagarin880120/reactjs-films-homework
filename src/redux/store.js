import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  searchResults: [],
  genres: [],
  isModalOpen: false,
  trailerURL: '',
  isTrailerLoaded: false,
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'SEARCH':
      return { ...state, searchResults: action.results };
    case 'GENRES':
      return { ...state, genres: action.genres };
    case 'MODAL':
      return { ...state, isModalOpen: action.isModalOpen };
    case 'TRAILER':
      return { ...state, trailerURL: action.trailerURL };
    case 'IS_TRAILER_LOADED':
      return { ...state, isTrailerLoaded: action.isTrailerLoaded };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
