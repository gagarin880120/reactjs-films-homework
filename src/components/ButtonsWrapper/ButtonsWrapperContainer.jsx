import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ButtonsWrapper from './ButtonsWrapper';
import { movieDetailsSelector } from '../../redux/selectors';
import { setModal, getTrailer } from '../../redux/actions';

export function ButtonsWrapperContainer({ movie, onTrailerButtonClick }) {
  return (
    <ButtonsWrapper
      movie={movie}
      onTrailerButtonClick={onTrailerButtonClick}
    />
  );
}

export const mapStateToProps = (state) => ({
  movie: movieDetailsSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onTrailerButtonClick(id) {
    dispatch(setModal(true));
    dispatch(getTrailer(id));
  },
});

ButtonsWrapperContainer.propTypes = {
  movie: PropTypes.instanceOf(Object),
  onTrailerButtonClick: PropTypes.func,
};

ButtonsWrapperContainer.defaultProps = {
  movie: {},
  onTrailerButtonClick: null,
};

export default connect(mapStateToProps, mapDispatchToProps)(ButtonsWrapperContainer);
