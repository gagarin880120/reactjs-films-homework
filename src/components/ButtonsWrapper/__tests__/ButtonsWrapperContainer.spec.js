import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import { ButtonsWrapperContainer, mapStateToProps, mapDispatchToProps } from '../ButtonsWrapperContainer';
import { setModal, getTrailer, getMovieDetails } from '../../../redux/actions';

jest.mock('../../../redux/actions',() => ({
  setModal: jest.fn().mockReturnValue(false),
  getTrailer: jest.fn().mockReturnValue('trailer'),
  getMovieDetails: jest.fn().mockReturnValue('details'),
}))

describe('ButtonsWrapperContainer', () => {
  describe('component', () => {
    test('should render without crashing', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<ButtonsWrapperContainer />);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    test('should return the right value', () => {

      const initialState = {
        movieDetails: {title: 'Matrix', runtime: 120, vote_average: 8.6, genres: [{name: 'Adventure'}]},
      };
      expect(mapStateToProps(initialState).movie).toEqual({
        genres: 'Adventure',
        runtime: '2h',
        stars: [
          {
            id: 0,
            type: 'full',
          },
          {
            id: 1,
            type: 'full',
          },
          {
            id: 2,
            type: 'full',
          },
          {
            id: 3,
            type: 'full',
          },
          {
            id: 4,
            type: 'half',
          },
        ],
        title: 'Matrix',
        vote_average: 8.6,
      });
    });
  });

  describe('mapDispatchToProps', () => {
    afterEach(() => {
      jest.clearAllMocks();
    })

    test('onTrailerButtonClick should dispatch setModal action', () => {
      const dispatch = jest.fn();
      const { onTrailerButtonClick } = mapDispatchToProps(dispatch);

      onTrailerButtonClick();

      expect(setModal).toHaveBeenCalled();
      expect(getTrailer).toHaveBeenCalled();
      expect(dispatch).toHaveBeenCalled();
    });
  })
})
