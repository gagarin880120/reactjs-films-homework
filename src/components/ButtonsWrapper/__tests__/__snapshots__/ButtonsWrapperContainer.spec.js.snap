// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonsWrapperContainer component should render without crashing 1`] = `
<ButtonsWrapper
  movie={Object {}}
  onTrailerButtonClick={null}
/>
`;
